src/ZEO/ClientStorage.py

```diff
@@ -559,7 +559,16 @@ def pack(self, t=None, referencesf=None, wait=1, days=0):
         if t is None:
             t = time.time()
         t = t - (days * 86400)
-        return self._call('pack', t, wait)
+        ret = self._call('pack', t, wait)
+        # remove all non-current entries from the cache.
+        # This way we make sure that loadBefore with before < packtime, won't
+        # return data from the cache, instead of returning "no data" if requested object
+        # has current revision >= packtime.
+        # By clearing all noncurrent entries we might remove more data from the
+        # cache than is strictly necessary, but since access to noncurrent data
+        # is seldom, that should not cause problems in practice.
+        self._cache.clearAllNonCurrent()
```

Comment:
Personally, I would like to see a specific direct test case for this.

While I know that there was a generic ZODB test failure, it's not obvious that this is the fix for it. What I mean is, the symptom (a failure to raise a particular exception) and the fix (making sure there are no non-current items in the cache) aren't necessarily clearly linked. I could imagine that a slight refactoring of the ZODB test that introduces additional clients or changes the timing or whatever might let it pass without this line of code being present.

I would suggest that a direct test for this would (a) establish non-current items in the cache and verify their existence in `cache.noncurrent`; (b) pack; (c) verify that `cache.noncurrent` is empty or equivalent.

Reply:
I agree, especially in the light of the bug you point out that the cache is not invalidated on other clients.